src/body/body.rs

```diff
@@ -326,21 +326,29 @@ impl Payload for Body {
         }
     }
 
-    fn content_length(&self) -> Option<u64> {
+    fn size_hint(&self) -> SizeHint {
         match self.kind {
-            Kind::Once(Some(ref val)) => Some(val.len() as u64),
-            Kind::Once(None) => Some(0),
-            Kind::Wrapped(..) => None,
-            Kind::Chan { content_length, .. } | Kind::H2 { content_length, .. } => content_length,
-        }
-    }
+            Kind::Once(Some(ref val)) => {
+                let mut hint = SizeHint::default();
+                hint.set_lower(val.len() as u64);
+                hint.set_upper(val.len() as u64);
+                hint
+            },
+            Kind::Once(None) => {
+                let mut hint = SizeHint::default();
+                hint.set_upper(0);
+                hint
+            },
+            Kind::Wrapped(..) => SizeHint::default(),
+            Kind::Chan { content_length, .. } | Kind::H2 { content_length, .. } => {
+                let mut hint = SizeHint::default();
 
-    // We can improve the performance of `Body` when we know it is a Once kind.
-    #[doc(hidden)]
-    fn __hyper_full_data(&mut self, _: FullDataArg) -> FullDataRet<Self::Data> {
-        match self.kind {
-            Kind::Once(ref mut val) => FullDataRet(val.take()),
-            _ => FullDataRet(None),
+                if let Some(content_length) = content_length {
+                    hint.set_upper(content_length as u64);
```

Comment:
Does this also need to `set_lower`? Otherwise the range is `0..content_length`, right?

Reply:
I think this to me makes the assumption that we have an upper bound but since the lower bound is not the same, poll_data will not return `0..upper`, whereas if `lower == upper` that means poll_data already has up too lower and upper and will return the full chunk.

Reply:
The intent of size_hint is to represent the total amount of data in the stream. If the total size of the stream is known, then lower should == upper, even if all the data is not received.

I have considered adding a field to indicate the amount of data that is ready to be consumed, but that has not been done yet.

Reply:
Exactly, so `SizeHint` means *total* data, not just next chunk. The `content-length` header is *total*.